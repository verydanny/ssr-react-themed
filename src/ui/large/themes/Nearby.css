@value ( Gray1 ) from 'ui/shared/themes/colors/colors.css';

.Nearby {
  display: flex;
  margin: 0 0 0 0;
}

.Nearby-leftColumn,
.Nearby-rightColumn {
  flex: 1;
}

.Nearby-leftColumn {
  padding-right: 25px;
}

.Nearby-leftColumn a,
.Nearby-rightColumn a {
  text-decoration: none;
  display: block;
  color: Gray1;
  font-size: 12px;
  font-weight: 600;
  line-height: 28px;
}
