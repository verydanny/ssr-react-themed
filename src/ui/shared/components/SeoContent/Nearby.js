import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'
import themed from 'react-themed-too'
import CollapsibleSection from './CollapsibleSection'
import ClampLines from './ClampLines'

@themed(/^Nearby/)

export default class Nearby extends PureComponent {
  static propTypes = {
    header: PropTypes.string.isRequired,
    linkTagSection: PropTypes.string.isRequired,
    linkTagItem: PropTypes.string.isRequired,
    data: PropTypes.array,
    maxLinks: PropTypes.number, // eslint-disable-line react/no-unused-prop-types
    theme: PropTypes.object,
  }

  static defaultProps = {
    data: [],
    theme: {},
  }

  constructor(props) {
    super(props)
    this.state = { columns: this.calcColumns(props) }
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.data !== this.props.data) {
      this.setState({ columns: this.calcColumns(nextProps) })
    }
  }

  sortDisplayName(a, b) {
    const nameA = a.displayName.toUpperCase()
    const nameB = b.displayName.toUpperCase()

    if (nameA < nameB) return -1
    if (nameA > nameB) return 1
    return 0
  }

  calcColumns(props) {
    const { data, maxLinks } = props

    if (!data || !data.length) return {}

    // links come in sorted by property count.
    // take the first <maxLinks> number of links before sorting alphabetically
    const links = maxLinks ? data.slice(0, maxLinks) : data
    const sorted = links.sort(this.sortDisplayName)
    const midpoint = Math.ceil(sorted.length / 2)
    const colLeft = sorted.slice(0, midpoint)
    const colRight = sorted.slice(midpoint)

    return { colLeft, colRight }
  }

  nearbyList() {
    const {
      linkTagSection,
      theme,
    } = this.props

    const {
      colLeft,
      colRight,
    } = this.state.columns

    if (!colLeft || !colLeft.length) return null

    return (
      <div
        className={theme.Nearby}
        data-tag_section={linkTagSection}
      >
        <div className={theme['Nearby-leftColumn']}>
          {this.renderColumn(colLeft)}
        </div>
        <div className={theme['Nearby-rightColumn']}>
          {this.renderColumn(colRight)}
        </div>
      </div>
    )
  }

  renderColumn(data = []) {
    const { linkTagItem } = this.props

    return data.map(({ url, displayName }, idx) => (
      <a
        key={`column-${idx}`}
        href={url}
        data-tag_item={linkTagItem}
      >
        {displayName}
      </a>
    ))
  }

  render() {
    const {
      header,
      data,
    } = this.props

    if (!data || !data.length) return null

    return (
      <CollapsibleSection
        header={header}
        isOpen
        testId="nearby-links"
      >
        <ClampLines
          lineCount={5}
          lineHeight={29}
        >
          {this.nearbyList()}
        </ClampLines>
      </CollapsibleSection>
    )
  }
}
