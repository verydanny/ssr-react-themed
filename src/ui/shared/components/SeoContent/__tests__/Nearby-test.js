import React from 'react'
import { mount } from 'enzyme'
import Nearby from '../Nearby'

const testLinks = [
  { displayName: 'Blandtown', url: '/georgia/atlanta-apartments/blandtown-neighborhood' },
  { displayName: 'Buckhead', url: '/georgia/atlanta-apartments/buckhead-neighborhood' },
  { displayName: 'Atkins Park', url: '/georgia/atlanta-apartments/atkins-park-neighborhood' },
  { displayName: 'Buford Highway', url: '/georgia/brookhaven-apartments/buford-highway-neighborhood' },
  { displayName: 'Capitol Gateway', url: '/georgia/atlanta-apartments/capitol-gateway-neighborhood' },
  { displayName: 'Chastain Park', url: '/georgia/atlanta-apartments/chastain-park-neighborhood' },
  { displayName: 'Virginia Highland', url: '/georgia/atlanta-apartments/virginia-highland-neighborhood' },
  { displayName: 'Colonial', url: '/georgia/atlanta-apartments/colonial-neighborhood' },
  { displayName: 'Westside', url: '/georgia/atlanta-apartments/westside-neighborhood' },
  { displayName: 'Eastside', url: '/georgia/atlanta-apartments/eastside-neighborhood' },
  { displayName: 'Downtown', url: '/georgia/atlanta-apartments/downtown-neighborhood' },
  { displayName: 'Lake Claire', url: '/georgia/atlanta-apartments/lake-claire-neighborhood' },
  { displayName: 'Brookhaven Fields', url: '/georgia/brookhaven-apartments/brookhaven-fields-neighborhood' },
  { displayName: 'Lenox', url: '/georgia/atlanta-apartments/lenox-neighborhood' },
  { displayName: 'Lindbergh - Morosgo', url: '/georgia/atlanta-apartments/lindbergh-morosgo-neighborhood' },
  { displayName: 'Kirkwood', url: '/georgia/atlanta-apartments/kirkwood-neighborhood' },
  { displayName: 'Midtown', url: '/georgia/atlanta-apartments/midtown-neighborhood' },
  { displayName: 'Morningside - Lenox Park', url: '/georgia/atlanta-apartments/morningside-lenox-park-neighborhood' },
  { displayName: 'Northyards', url: '/georgia/atlanta-apartments/northyards-neighborhood' },
  { displayName: 'Oakland', url: '/georgia/atlanta-apartments/oakland-neighborhood' },
  { displayName: 'North Druid Woods', url: '/georgia/north-decatur-apartments/north-druid-woods-neighborhood' },
  { displayName: 'Pine Hills', url: '/georgia/atlanta-apartments/pine-hills-neighborhood' },
  { displayName: 'Peachtree Battle', url: '/georgia/atlanta-apartments/peachtree-battle-neighborhood' },
  { displayName: 'Piedmont Heights', url: '/georgia/atlanta-apartments/piedmont-heights-neighborhood' },
  { displayName: 'Northeast Atlanta', url: '/georgia/atlanta-apartments/northeast-atlanta-neighborhood' },
  { displayName: 'English Avenue', url: '/georgia/atlanta-apartments/english-avenue-neighborhood' },
  { displayName: 'Sherwood Forest', url: '/georgia/atlanta-apartments/sherwood-forest-neighborhood' },
  { displayName: 'South Downtown', url: '/georgia/atlanta-apartments/south-downtown-neighborhood' },
  { displayName: 'Vine City', url: '/georgia/atlanta-apartments/vine-city-neighborhood' },
  { displayName: 'West Paces Ferry', url: '/georgia/atlanta-apartments/west-paces-ferry-neighborhood' },
]

const theme = {
  'Nearby-leftColumn': 'leftColumn',
  'Nearby-rightColumn': 'rightColumn',
}

const props = {
  header: 'Nearby Header',
  linkTagSection: 'tag_section',
  linkTagItem: 'tag_item',
  data: testLinks,
  theme,
}

describe('ui/shared/components/SeoContent/Nearby', () => {
  it('matches snapshot', () => {
    const wrapper = mount(<Nearby {...props} />)
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('sets a header', () => {
    const wrapper = mount(<Nearby {...props} />)
    const header = wrapper.find('h2')
    expect(header.text()).toBe(props.header)
  })

  it('has a left column of links', () => {
    const wrapper = mount(<Nearby {...props} />)
    const leftColumn = wrapper.find('.leftColumn')
    expect(leftColumn.length).toBe(1)
    const links = leftColumn.find('a')
    expect(links.length).toBe(15)
    expect(links.at(0).text()).toBe('Atkins Park')
  })

  it('has a right column of links', () => {
    const wrapper = mount(<Nearby {...props} />)
    const rightColumn = wrapper.find('.rightColumn')
    expect(rightColumn.length).toBe(1)
    const links = rightColumn.find('a')
    expect(links.length).toBe(15)
    expect(links.at(0).text()).toBe('Midtown')
  })

  it('can limit the number of links', () => {
    const wrapper = mount(<Nearby {...props} maxLinks={10} />)
    const leftColumn = wrapper.find('.leftColumn')
    expect(leftColumn.length).toBe(1)
    let links = leftColumn.find('a')
    expect(links.length).toBe(5)
    expect(links.at(0).text()).toBe('Atkins Park')

    const rightColumn = wrapper.find('.rightColumn')
    expect(rightColumn.length).toBe(1)
    links = rightColumn.find('a')
    expect(links.length).toBe(5)
    expect(links.at(0).text()).toBe('Capitol Gateway')
  })
})
