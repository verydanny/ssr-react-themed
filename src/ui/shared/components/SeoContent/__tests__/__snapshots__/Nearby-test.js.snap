// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ui/shared/components/SeoContent/Nearby matches snapshot 1`] = `

<div data-tid="nearby-links">
  <h2 data-tag_item="header">
    <button>
      Nearby Header
    </button>
  </h2>
  <div>
    <div>
      <div data-tag_section="tag_section">
        <div class="leftColumn">
          <a href="/georgia/atlanta-apartments/atkins-park-neighborhood"
             data-tag_item="tag_item"
          >
            Atkins Park
          </a>
          <a href="/georgia/atlanta-apartments/blandtown-neighborhood"
             data-tag_item="tag_item"
          >
            Blandtown
          </a>
          <a href="/georgia/brookhaven-apartments/brookhaven-fields-neighborhood"
             data-tag_item="tag_item"
          >
            Brookhaven Fields
          </a>
          <a href="/georgia/atlanta-apartments/buckhead-neighborhood"
             data-tag_item="tag_item"
          >
            Buckhead
          </a>
          <a href="/georgia/brookhaven-apartments/buford-highway-neighborhood"
             data-tag_item="tag_item"
          >
            Buford Highway
          </a>
          <a href="/georgia/atlanta-apartments/capitol-gateway-neighborhood"
             data-tag_item="tag_item"
          >
            Capitol Gateway
          </a>
          <a href="/georgia/atlanta-apartments/chastain-park-neighborhood"
             data-tag_item="tag_item"
          >
            Chastain Park
          </a>
          <a href="/georgia/atlanta-apartments/colonial-neighborhood"
             data-tag_item="tag_item"
          >
            Colonial
          </a>
          <a href="/georgia/atlanta-apartments/downtown-neighborhood"
             data-tag_item="tag_item"
          >
            Downtown
          </a>
          <a href="/georgia/atlanta-apartments/eastside-neighborhood"
             data-tag_item="tag_item"
          >
            Eastside
          </a>
          <a href="/georgia/atlanta-apartments/english-avenue-neighborhood"
             data-tag_item="tag_item"
          >
            English Avenue
          </a>
          <a href="/georgia/atlanta-apartments/kirkwood-neighborhood"
             data-tag_item="tag_item"
          >
            Kirkwood
          </a>
          <a href="/georgia/atlanta-apartments/lake-claire-neighborhood"
             data-tag_item="tag_item"
          >
            Lake Claire
          </a>
          <a href="/georgia/atlanta-apartments/lenox-neighborhood"
             data-tag_item="tag_item"
          >
            Lenox
          </a>
          <a href="/georgia/atlanta-apartments/lindbergh-morosgo-neighborhood"
             data-tag_item="tag_item"
          >
            Lindbergh - Morosgo
          </a>
        </div>
        <div class="rightColumn">
          <a href="/georgia/atlanta-apartments/midtown-neighborhood"
             data-tag_item="tag_item"
          >
            Midtown
          </a>
          <a href="/georgia/atlanta-apartments/morningside-lenox-park-neighborhood"
             data-tag_item="tag_item"
          >
            Morningside - Lenox Park
          </a>
          <a href="/georgia/north-decatur-apartments/north-druid-woods-neighborhood"
             data-tag_item="tag_item"
          >
            North Druid Woods
          </a>
          <a href="/georgia/atlanta-apartments/northeast-atlanta-neighborhood"
             data-tag_item="tag_item"
          >
            Northeast Atlanta
          </a>
          <a href="/georgia/atlanta-apartments/northyards-neighborhood"
             data-tag_item="tag_item"
          >
            Northyards
          </a>
          <a href="/georgia/atlanta-apartments/oakland-neighborhood"
             data-tag_item="tag_item"
          >
            Oakland
          </a>
          <a href="/georgia/atlanta-apartments/peachtree-battle-neighborhood"
             data-tag_item="tag_item"
          >
            Peachtree Battle
          </a>
          <a href="/georgia/atlanta-apartments/piedmont-heights-neighborhood"
             data-tag_item="tag_item"
          >
            Piedmont Heights
          </a>
          <a href="/georgia/atlanta-apartments/pine-hills-neighborhood"
             data-tag_item="tag_item"
          >
            Pine Hills
          </a>
          <a href="/georgia/atlanta-apartments/sherwood-forest-neighborhood"
             data-tag_item="tag_item"
          >
            Sherwood Forest
          </a>
          <a href="/georgia/atlanta-apartments/south-downtown-neighborhood"
             data-tag_item="tag_item"
          >
            South Downtown
          </a>
          <a href="/georgia/atlanta-apartments/vine-city-neighborhood"
             data-tag_item="tag_item"
          >
            Vine City
          </a>
          <a href="/georgia/atlanta-apartments/virginia-highland-neighborhood"
             data-tag_item="tag_item"
          >
            Virginia Highland
          </a>
          <a href="/georgia/atlanta-apartments/west-paces-ferry-neighborhood"
             data-tag_item="tag_item"
          >
            West Paces Ferry
          </a>
          <a href="/georgia/atlanta-apartments/westside-neighborhood"
             data-tag_item="tag_item"
          >
            Westside
          </a>
        </div>
      </div>
    </div>
  </div>
</div>

`;
